Remove `Internal` from `gh repo create` prompt when owner is not an org

jtmcg opened the pull request:
Fixes #9464

Internal repos only exist for organizations, so when a user selects their personal namespace to create a repo using `gh repo create`, `Internal` should not be an option in the `Visibility` prompt.

This should avoid the additional quirk where if the user selects `Internal` while creating a personal repo and then proceeds to add any of the README, .gitignore, or LICENSE files prompted for later, the repo will not error and instead get created as a `Public` repo. This has the potential for a user to unknowingly leak sensitive info intended to go into a non-public repo.

## Screenshots

### Visibility prompt when creating a personal repo

<img width="617" alt="Screenshot 2024-08-14 at 3 26 37 PM" src="https://github.com/user-attachments/assets/2a263826-e7e6-4721-8a11-f760a5f5f29a">

### Visibility prompt when creating an org repo

<img width="620" alt="Screenshot 2024-08-14 at 3 26 14 PM" src="https://github.com/user-attachments/assets/3e93dc7b-f058-46d2-8e45-9fc40d8cdebc">

## To test

1. Pull down branch `gh pr checkout 9465`
2. Build the tool in the root directory: `make`
3. Try the prompt, selecting your personal account: `bin/gh repo create`

jtmcg committed:
Remove `Internal` from `gh repo create` prompt when owner is not an org

Closes #9464

Internal repos only exist for organizations, so when a user selects their
personal namespace to create a repo using `gh repo create`, `Internal`
should not be an option in the `Visibility` prompt.

This should avoid the additional quirk where if the user selects
`Internal` while creating a personal repo and then proceeds to add any
of the README, .gitignore, or LICENSE files prompted for later, the repo
will not error and instead get created as a `Public` repo. This has the
potential for a user to unknowingly leak sensitive info intended to go
into a non-public repo.

andyfeller commented on pkg/cmd/repo/create/create.go:
```diff
@@ -855,7 +855,11 @@ func interactiveRepoInfo(client *http.Client, hostname string, prompter iprompte
 	}
 
-	visibilityOptions := []string{"Public", "Private", "Internal"}
+	visibilityOptions := []string{"Public", "Private"}
+	// orgs can also create internal repos
+	if owner != "" {
+		visibilityOptions = append(visibilityOptions, "Internal")
+	}
```
This change makes sense after reviewing `interactiveRepoNameAndOwner` will return a non-empty string for owner if an organization:

https://github.com/cli/cli/blob/74b9724c02876b024d03f4c26680ed0ed79860ac/pkg/cmd/repo/create/create.go#L903-L908

Outside the scope of this issue, there are some types of organizations that cannot create `public` repositories like [Enterprise Managed Users](https://docs.github.com/en/enterprise-cloud@latest/admin/managing-iam/understanding-iam-for-enterprises/about-enterprise-managed-users); only noting a similar use case which would require inspection of the selected owner to detect.

> Managed user accounts cannot create public content or collaborate outside your enterprise. See "[Abilities and restrictions of managed user accounts](https://docs.github.com/en/enterprise-cloud@latest/admin/identity-and-access-management/understanding-iam-for-enterprises/abilities-and-restrictions-of-managed-user-accounts)."

andyfeller commented on pkg/cmd/repo/create/create.go:
```diff
@@ -855,7 +855,11 @@ func interactiveRepoInfo(client *http.Client, hostname string, prompter iprompte
 	}
 
-	visibilityOptions := []string{"Public", "Private", "Internal"}
+	visibilityOptions := []string{"Public", "Private"}
+	// orgs can also create internal repos
+	if owner != "" {
+		visibilityOptions = append(visibilityOptions, "Internal")
+	}
```
How might we enhance tests to account for this change?

https://github.com/cli/cli/blob/74b9724c02876b024d03f4c26680ed0ed79860ac/pkg/cmd/repo/create/create_test.go

jtmcg replied:
No tests currently exist for the function `interactiveRepoInfo`. I could certainly add tests to cover it, if so desired. I do not think testing this through the addition of tests in `Test_createRun` is prudent.

My preference might be to just ship this then refactor this command to support better coverage of these kinds of changes. What do you think?

williammartin replied:
Talk to me more about why you wouldn't want to test this through `Test_createRun`?

jtmcg replied:
Good question 🤔  In addition to this proliferating of a bad testing pattern (as mentioned above), as far as I can tell, the prompter tests are incapable of testing what is being prompted (in this test suite) without [mocking out incorrect user input](https://github.com/cli/cli/blob/4283adbc40d1f0ebad6728feffe30f5a46105140/pkg/cmd/repo/create/create_test.go#L414). Since the user isn't able to select a prompt that doesn't exist, we aren't able to actually detect changes to what is prompted - only testing errors in user input for the prompter. That, to me, smells like over-testing, as the [prompter already covers that use case](https://github.com/AlecAivazis/survey/blob/160123ed7175f77bc8625b59c44b8dc03f5333e3/select_test.go#L308-L328). 

All these taken together means, to me, that the logic we're _actually_ interested in testing is how `visibilityOptions` responds to a change in the "owner" variable (return value from the api call). I suppose we could include a test for that, which results in a tiny refactor. I've [committed](https://github.com/cli/cli/pull/9465/commits/3c19d28df59d4dea27596d65f4c547f4f96906e3) what that looks like for comparison, here.

Final point: making this change didn't cause any failures to begin with, and given there is the potential for users to unknowingly do something dangerous (set a repo to "Public" when they intended "Private/Internal") with the current behavior I thought that shipping this in favor of a refactor was prudent.

Let me know what you think about the tested commit option I've included 🙂

jtmcg replied:
> Outside the scope of this issue, there are some types of organizations that cannot create public repositories like [Enterprise Managed Users](https://docs.github.com/en/enterprise-cloud@latest/admin/managing-iam/understanding-iam-for-enterprises/about-enterprise-managed-users); only noting a similar use case which would require inspection of the selected owner to detect.

A note, @andyfeller, is that this small change probably makes addressing the above easier as well 🙌

williammartin replied:
My concern about testing the private function with the intention to refactor in the future is that we don't retain any regression safety. However, I think that refactoring this command is such a large piece of work, requiring so much tearing out that the entire test suite will need to be rewritten anyway. I'm content to move forward with your change and to add these tests to the existing table tests because frankly, the existing pattern is extremely annoying.

> Since the user isn't able to select a prompt that doesn't exist, we aren't able to actually detect changes to what is prompted

If I understand you correctly, I do think that we can assert on the `options` inside the prompter test doubled `SelectFunc ` functions e.g.:

```go
p.SelectFunc = func(message, defaultValue string, options []string) (int, error) {
	switch message {
...
	case "Visibility":
		if slices.Contains(options, "Internal") {
			return 0, fmt.Errorf("should not contain internal visibility option: %v", options)
		}
		return prompter.IndexFor(options, "Private")
...
        }
}
```

But I'm only mentioning this for interest sake. I don't really want to write this test because it's a massive amount of noise.

jtmcg committed:
Refactor prompter with test coverage

By extracting the repo visibility options to its own function,
getRepoVisibilityOptions, we're able to directly test the behavior
introduced with this change. This breaks the testing pattern established
here thus far, but may be a good example of the direction we should
explore for a future refactor.

williammartin commented:
Bringing a convo from private to here...

Although this PR currently avoids accidentally creating an internal repo as a private one in the interactive flow, it is still possible to accidentally do this in a non-interactive flow:

Both of these are from your branch:

First one shows Internal not in prompt

```
➜  cli git:(fix-prompting-for-gh-repo-create) ./bin/gh repo create
? What would you like to do? Create a new repository on GitHub from scratch
? Repository name visibility-triage-2
? Repository owner williammartin
? Description
? Visibility  [Use arrows to move, type to filter]
> Public
  Private
```

Second one shows internal flag resulting in this being created publically

```
➜  cli git:(fix-prompting-for-gh-repo-create) ./bin/gh repo create visibility-triage-2 --add-readme --internal
✓ Created repository williammartin/visibility-triage-2 on GitHub
  https://github.com/williammartin/visibility-triage-2
```

Just for evidence that the org cannot have internal (when I forgot to add the readme):

```
➜  cli git:(fix-prompting-for-gh-repo-create) ./bin/gh repo create visibility-triage-2 --internal
GraphQL: Only organization-owned repositories can have internal visibility (createRepository)
```

I think probably the right thing to do is like…. right here: https://github.com/cli/cli/blob/3c19d28df59d4dea27596d65f4c547f4f96906e3/pkg/cmd/repo/create/http.go#L185

```
isInternal := strings.ToLower(input.Visibility) == "internal"
if isInternal && !isOrg {
  return nil, fmt.Errorf("blaghdgasghas")
}
```

jtmcg committed:
Add failing tests to check for error with internal vis in non-org repos

There is a bug in the code, currently, where a user repo can attempt to be
created as with `--internal` visibility flag  when that is not an option
for non-org repos. It fails at the API level if the --gitignore,
--license, or --add-readme flags are not included, but silently falls back
to Public visibility if one of them is included.

Because this bug already existed, this commit adds the tests to ensure
that both scenarios described above are captured accurately by the test
suite. A fix for the latter scenario will be coming in a future commit

jtmcg committed:
Add Exclude to httpmock registry and implement in Test_repoCreate

Upon attempting to make the previous commit pass, I realized that it was
actually impossible to test what I wanted to. The tests in the previous
commit were behaving as expected given the bug that commit described, but
upon attempting to implement a solution I realized that the tests were
only testing the mocks and not the code functionality itself.

Essentially, when the code to fix the bug was implemented, the tests were
failing because the mocks required to test the buggy behavior were no
longer being called. To make the tests pass, I'd have to rewrite them, but
were I to remove the bug fix, the tests would no longer fail.

This pointed me to a gap in our httpmocks - the ability to intentionally
exclude api calls. The behavior I'm trying to test, here, is that we stop
executing when a certain condition is met, and therefore won't make any
subsequent api calls down the chain.

This implements the Exclude method on the registry such that it will fail
if an excluded api pattern is called. I have refactored the tests in
Test_repoCreate to use the Exclude mock for testing.

jtmcg committed:
Add error if user attempts to create repo with --internal flag

This was previously failing at either the API if no other flags were
included or falling back to creating a public repo if one of gitignore,
license, or add-readme were included.

andyfeller commented on pkg/httpmock/registry.go:
```diff
@@ -25,6 +28,18 @@ func (r *Registry) Register(m Matcher, resp Responder) {
 	})
 }
 
+func (r *Registry) Exclude(t *testing.T, m Matcher) {
+	excludedStub := &Stub{
+		Matcher: m,
+		Responder: func(req *http.Request) (*http.Response, error) {
+			assert.FailNowf(t, "Exclude error", "API called when excluded: %v", req.URL)
+			return nil, nil
+		},
+		exclude: true,
+	}
+	r.stubs = append(r.stubs, excludedStub)
+}
+
```
**How readily can this new capability be reasoned in context of `httpmock`?**

This is the question I mentally keep coming back to reading and re-reading through this updated `registry` code.  Depending how accurate I am below, maybe some added documentation and/or renaming makes sense:

- Tests that `Register` stubs are ensuring that specific HTTP requests are being made for REST or GraphQL APIs when the test calls `Verify` after being deferred.  It in turn iterates over all the stubs registered, finds whether any stubs have not been matched, and errors appropriately.

- `Exclude` now changes this dynamic by including stubs that cause the test to fail if encountered.  Less of a `Exclude` and more of a `FailIf` scenario in my mind.

- Lastly, I don't know if `exclude` field is actually needed if an `Exclude`ed stub will be marked as `matched` in `RoundTrip`:

https://github.com/cli/cli/blob/95b83872156a7dd2f16cd37ad7b29de2871a2d85/pkg/httpmock/registry.go#L64-L83

jtmcg replied:
> Lastly, I don't know if exclude field is actually needed if an Excludeed stub will be marked as matched in RoundTrip

Unfortunately this is required because `Verify` will complain for any unmatched stubs. 

```
$ go test
--- FAIL: Test_repoCreate (0.00s)
    --- FAIL: Test_repoCreate/create_personal_repository_but_try_to_set_it_as_'internal' (0.00s)
        http_test.go:737: 1 unmatched HTTP stubs
    --- FAIL: Test_repoCreate/create_personal_repository_with_README_but_try_to_set_it_as_'internal' (0.00s)
        http_test.go:737: 1 unmatched HTTP stubs
FAIL
exit status 1
FAIL    github.com/cli/cli/v2/pkg/cmd/repo/create       3.247s
```

There's another implementation I tried that adds an `excludedStubs []stub` key/value to the `Registry` struct, but I tinkered with that for a long time and the looping and exiting of the tests weren't very user friendly. There ended up being extra log dump to parse for the Exclusion failure that was bad UX and the `exclude` field is where I landed.

> Exclude now changes this dynamic by including stubs that cause the test to fail if encountered. Less of a Exclude and more of a FailIf scenario in my mind

First, names are hard 😅 I'd love to figure out what best to name this 📛 

That said, I think `Exclude` isn't bad because technically we're calling `registry.Exclude` when we use this - it's just shortened to `r.Exclude` in many of these tests (it can be confusing because we're shortening the name of the registry. Shortening the names of receivers and variables in golang is one of the things I take exception to in Golang practices for precisely this reason). So really our interface is:

```Golang
type Registry interface {
    Register(Matcher, Responder)
    Exclude(*testing.T, Matcher)
    Verify(Testing)
    RoundTrip(*http.Request)
}
```

Taken all together, you either "register" and api call "exclude" an api calls. 

I'm not sure if `FailIf` fits that well. I'm open to other suggestions.

andyfeller replied:
> First, names are hard 😅

Yes they are!

Putting aside whether to change the name of this or not, **I primarily wanted to make sure I was following what this intended**.

Because there is little to no documentation around the `httpmock` package, it causes the reader to read the code and infer the intention rather than clearly calling out what it does with documentation.  So I'm making sure I understand its use first 👍 

My question about whether the `exclude` field was needed or not was because it seemed like the excluded stubs were still marked as matched, but I trust you; just unclear how that 1 additional field changes anything assuming everything is marked matched.

jtmcg replied:
Ah, gotcha. Yeah, your understanding is correct, but I'll add one piece that I think will help solidify this

> Tests that Register stubs are ensuring that specific HTTP requests are being made for REST or GraphQL APIs when the test calls Verify after being deferred. It in turn iterates over all the stubs registered, finds whether any stubs have not been matched, and errors appropriately.

When a match is found (and before anything errors), it also runs the [provided callback, called `Responder`](https://github.com/cli/cli/blob/c81ccab4b8bb1ef1c774f4ba9851e9dcb11fa5e7/pkg/httpmock/registry.go#L77), on the stub. I've leveraged this so that the callback is hardcoded for the `Exclude` method so that when the match is found the callback that is invoked results in a failing test. That way I can avoid relying on errors to communicate that an excluded endpoint was called. This is important because testing for errors is a common pattern throughout the code (see [wantErr](https://github.com/cli/cli/blob/c81ccab4b8bb1ef1c774f4ba9851e9dcb11fa5e7/pkg/cmd/repo/create/create_test.go#L858-L862))

andyfeller reviewed (approved):
Putting aside any significant improvements to our testing approach around the code under review, this seems solid ✨.

I'm unsure if you're planning to incorporate any more changes into this, so will keep an eye open.

jtmcg committed:
Add testing for error messages in gh repo create

In the previous commits, we've introduced a new error when a user tries to
create an Internal repo not owned by an organization. This adds tests to
verify that the error we are getting is, in fact, the one associated with
this use case and not some random error.

jtmcg commented:
Ugh, I borked it somehow... Fixing git history

jtmcg commented:
> Ugh, I borked it somehow... Fixing git history

Fixed now. @andyfeller, would you mind giving one more cursory 👀 to make sure I didn't miss anything with `git` shenanigans? Note - I also added tests checking the error message as well

williammartin reviewed (approved):
LGTM thanks!
